### DISCUSSION TITLE
Union of dicts with heterogeneous value types

### DISCUSSION THREAD

**Discussion by harahu** [python/typing#1895] [Q&A] [answered]
I'm running into a type checking error where I'm unsure whether what I'm looking at is an issue in mypy or an issue in my understanding of python type annotations.

See: https://mypy-play.net/?mypy=master&python=3.13&flags=show-error-codes&gist=2643cb3ff1ea956a1c4ed0869140d46d

```
d1: dict[str, str] = {}
d2: dict[str, str | None] = {}

d3 = d1 | d2
reveal_type(d3)  # builtins.dict[builtins.str, Union[builtins.str, None]], aka dict[str, str | None]

# This is fine
d4: dict[str, str | None] = d3  # aka d1 | d2

# This isn't?
d5: dict[str, str | None] = d1 | d2
```

I cannot figure out why I'm not allowed to do `d5: dict[str, str | None] = d1 | d2` in this case.

**Comment by erictraut** [comment, marked as answer]
I think this is an issue with mypy. The code looks fine to me, and pyright accepts it [without any errors](https://pyright-play.net/?pyrightVersion=1.1.389&pythonVersion=3.13&code=CYRgXABMCWDGAuBtAzvATgGgqtBdCAvBAN4C%2BAUMAEyQwIrpY4QA%2BEAcgPYB2ApvkTLlKAZkJQQrKFXJpeAN14BDADYB9eAE8ADrwAUwEQEoIEAMQQARgFdoK%2BNG7IAdHSQ27Dp85xYAqtzQPIge9o4uvhw8-LhYSgDWSlBwSJHMbFx8uMIWACoAFtDIEEUQAGaOvJQALLQpDJjY6FKZ-OKGphYJSaBS1DkQBaVF3ADk8AD8lACsdfRpzRnRAhJ9MkA).
